### PR TITLE
Add support for temporaries

### PR DESCRIPTION
**Pull request by GuntramStrecha**
- Adjust edge cases and mismatch
- Add temporaries coverage
- Rework iterator/reference types
- Add iterator benchmarks for different container sizes

### CODE REVIEW

File: tests/temp.cpp [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunsafe-buffer-usage"
+#include <catch2/catch_test_macros.hpp>
+#pragma clang diagnostic pop
+#else
+#include <catch2/catch_test_macros.hpp>
+#endif
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-member-function"
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply tests/temp.cpp]: Fixed

File: tests/temp.cpp [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+    THEN("all_of works")
+    {
+      auto const all_positive = std::ranges::all_of(
+        gst::ranges::views::zip({1, 2, 3}, {'a', 'b', 'c'}),
+        [](auto const& t) { return std::get<0>(t) > 0 && std::get<1>(t) != '\0'; });
+      REQUIRE(all_positive);
+    }
+
+    THEN("accumulate works")
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply tests/temp.cpp]: Fixed.

File: tests/temp.cpp [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    auto const std_zip = std::ranges::views::zip(std::array<int, 0>{}, std::list<char>{});
+
+    THEN("The zip_view should be empty") { REQUIRE(gst_zip.empty() == std_zip.empty()); }
+    THEN("The size should be zero") { REQUIRE(gst_zip.size() == std_zip.size()); }
+  }
+}
+
+SCENARIO("Testing gst::ranges::zip_view with temporaries", "[temporaries]")
+{
+  GIVEN("a zip_view created from a temporary container")
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply tests/temp.cpp]: Fixed.

File: tests/temp.cpp [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+  {
+    auto const gst_zip =
+      gst::ranges::views::zip(std::array<int, 3>{1, 2, 3}, std::list<char>{'a', 'b'});
+    auto const std_zip =
+      std::ranges::views::zip(std::array<int, 3>{1, 2, 3}, std::list<char>{'a', 'b'});
+
+    THEN("The zip_view should be empty") { REQUIRE(gst_zip.empty() == std_zip.empty()); }
+    THEN("The size should be zero") { REQUIRE(gst_zip.size() == std_zip.size()); }
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply tests/temp.cpp]: Fixed.

File: include/zip_view.hpp [hunk 5]
```diff
@@ -87,10 +88,24 @@ struct zip_iterator_category_impl<Iter, Rest...>
...
+auto as_lvalue(store_t<T>& t) noexcept -> store_t<T>&
+{
+  return t;
+}
+
+template <typename T, typename std::enable_if<!std::is_reference<store_t<T>>::value, int>::type = 0>
+auto as_lvalue(store_t<T> const& t) noexcept -> store_t<T> const&
+{
+  return t;
+}
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply include/zip_view.hpp]: Fixed.

File: include/zip_view.hpp [hunk 6]
```diff
@@ -372,26 +431,20 @@ class zip_view
...
-template <typename... Containers>
-auto zip(Containers&&... containers) -> zip_view<Containers...>
+template <typename... Ts>
+auto zip(std::initializer_list<Ts>&&... initializer_list)
+  -> zip_view<std::initializer_list<Ts>&&...>
 {
-  return zip_view<Containers...>(std::forward<Containers>(containers)...);
+  return zip_view<std::initializer_list<Ts>&&...>(
+    std::forward<std::initializer_list<Ts>>(initializer_list)...);
 }
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply include/zip_view.hpp]: Remove support for rvalue initializer_lists.

File: include/zip_view.hpp [hunk 7]
```diff
@@ -352,17 +411,17 @@ class zip_view
...
   }
 
-  auto           begin() const -> iterator { return iterator(begin_impl(INDICES)); }
+  auto           begin() const -> const_iterator { return const_iterator(begin_impl(INDICES)); }
   auto           begin() -> iterator { return iterator(begin_impl(INDICES)); }
-  auto           end() const -> iterator { return iterator(end_impl(INDICES)); }
-  auto           end() -> iterator { return iterator(end_impl(INDICES)); }
+  auto           end() const -> const_iterator { return const_iterator(cached_end_const_); }
+  auto           end() -> iterator { return iterator(cached_end_mut_); }
   constexpr auto size() const -> std::size_t { return min_size(INDICES); }
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply include/zip_view.hpp:499]: Caching removed.

File: tests/temp.cpp [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+      STATIC_REQUIRE(std::is_lvalue_reference<gst_elem_ref_t>::value ==
+                     std::is_lvalue_reference<std_elem_ref_t>::value);
+      STATIC_REQUIRE(std::is_const<std::remove_reference_t<gst_elem_ref_t>>::value ==
+                     std::is_const<std::remove_reference_t<std_elem_ref_t>>::value);
+    }
+
+    THEN("can mutate elements through const zip_view")
+    {
+      REQUIRE(std::get<0>(gst_zip[0]) == std::get<0>(std_zip[0]));
+    }
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply tests/temp.cpp]: Fixed.

File: include/zip_view.hpp [hunk 9]
```diff
@@ -325,11 +370,25 @@ class zip_view
...
-  zip_view(zip_view const& other) : containers_(other.containers_) {}
-  zip_view(zip_view&& other) noexcept : containers_(std::move(other.containers_)) {}
+  using iterator       = basic_iterator<iterators_mut>;
+  using const_iterator = basic_iterator<iterators_const>;
+
+  template <typename... Cs,
+            typename std::enable_if<sizeof...(Cs) == sizeof...(Containers), int>::type = 0>
+  explicit zip_view(Cs&&... containers)
+    : containers_(std::forward<Cs>(containers)...), cached_size_(min_size(INDICES)),
+      cached_end_mut_(end_impl(INDICES)), cached_end_const_(end_impl(INDICES))
```
[the remark this answers is not in the document]
Comment (GuntramStrecha) [reply include/zip_view.hpp]: Caching removed.
